Automattic/wp-calypso
# Editor: Move nested sidebar state to redux

### Summary
The PostEditor and it's child components are laden with extraneously passed props and state, causing a lot of tight coupling and rigidity.
This piece of work aims to alleviate that problem a little and take steps towards making PostEditor and the other components nested within a little easier to deal with.

### Testing
There should be no behavioural changes resulting from this change-set, everything should work as before.
Because there are a lot of areas effected by these changes, there is a fair amount to test - please bare with me...

*Note*: We'll need to be sure that no analytics are lost in these changes, so before you start testing set analytics debugging on with: `localStorage.setItem('debug', 'calypso:analytics:*');`

#### Testing the EditorGroundControl 'history' button
- Starting with a closed sidebar...
  - Click the 'history' button.
  - Note that the sidebar opens, showing the revisions list.
    - Tracking: Note that 'calypso_editor_post_revisions_open' tracks event fires
    - Tracking: Note that 'editor_actions:open-sidebar' mc stat is bumped
  - Click the history button again.
  - Note that the revisions list is hidden, leaving the sidebar open.
- Starting with an open sidebar...
  - Click the 'history' button.
  - Note that revisions list is shown in the sidebar.
    - Tracking: Note that 'calypso_editor_post_revisions_open' tracks event fires
  - Click the history button again
  - Note that the revisions list is hidden, leaving the sidebar open.

#### Testing the 'Post Settings' action button
- Click this button repeatedly
- Note that each time the sidebar 'toggles' between open & closed states.
    - Tracking: Note that on opening 'editor_actions:open-sidebar' mc stat is bumped
    - Tracking: Note that on closing 'editor_actions:close-sidebar' mc stat is bumped

#### Testing the 'xx Revisions' button
- Make sure the sidebar is open & the 'status' drawer is expanded
- Click the 'xx Revisions' button (noted in the below image).
- Note that revisions list is shown in the sidebar.
    - Tracking: Note that 'calypso_editor_post_revisions_open' tracks event fires

<img width="321" alt="screen shot 2017-11-07 at 23 28 19" src="https://user-images.githubusercontent.com/4335450/32523531-94821d56-c413-11e7-84d7-561af9590c5f.png">


#### Testing the 'X' (close sidebar) button
- Starting with a closed sidebar...
  - Open the sidebar using the 'Post Settings' button.
  - Look for the 'X' button in the sidebar and press it.
  - Note that the sidebar closes.
    - Tracking: Note that 'editor_actions:close-sidebar' mc stat is bumped
- Starting with a closed sidebar...
  - Click the 'history' button (editor ground control header) to open the revisions list in the sidebar
  - Look for the 'X' button in the sidebar and press it. 
  - Note that the revisions list closes and leaves the sidebar open.
  - Click the 'X' button again. 
  - Note that the sidebar closes.
    - Tracking: Note that 'editor_actions:close-sidebar' mc stat is bumped

### Follow-up Work
{ I'll add tasks here that I feel should be worked on relating to these changes but that are outside the scope of this already large change-set }

- Connect `isSidebarOpened` prop in  `HistoryButton`
- When clicking the history button, is there a nice way to hide the sidebar if it was originally opened with that button? (This is the behaviour I would expect)
- Create a `goBack` method on `editor-ground-control` to avoid binding in the render method.
- Remove the need to call `selectRevision( null )` when HistoryButton opens the NestedSidebar there's a more logical event for reseting the selection.
- Componentise the  'toggle' action button and connect the `toggleSidebar` method, avoiding the need to pass it as a prop through `EditorGroundControl`.
